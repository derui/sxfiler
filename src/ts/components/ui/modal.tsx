import { h } from "preact";
import { createPortal } from "preact/compat";
import classnames from "classnames";

type BaseProps = preact.JSX.HTMLAttributes<HTMLElement> & {
  position?: "top" | undefined;
};

type BaseContainerProps = {
  opened: boolean;
};

type BaseOverlayProps = {
  opened: boolean;
};

export type Props<ContainerProps extends object = {}, T extends BaseProps = BaseProps> = T & {
  dialogRoot: HTMLElement;
  opened: boolean;
  container?: ContainerProps;
};

const DefaultRoot: preact.FunctionComponent<BaseProps> = (props) => {
  const className = classnames({
    "ui-modal__root": !props.position,
    "ui-modal__root--top": props.position === "top",
  });

  return <div class={className} data-testid="modal-root" {...props} />;
};
DefaultRoot.displayName = "DefaultRoot";

const Overlay: preact.FunctionComponent<BaseOverlayProps> = ({ opened }) =>
  opened ? <div class="ui-modal__overlay" data-testid="modal-overlay" /> : null;
Overlay.displayName = "Overlay";

export const createComponent = function createComponent<
  ContainerProps extends object = {},
  T extends BaseProps = BaseProps
>(context: {
  root?: preact.ComponentType<BaseProps>;
  container: preact.ComponentType<ContainerProps & BaseContainerProps>;
}): preact.ComponentType<Props<ContainerProps, T>> {
  const Root = context.root || DefaultRoot;
  const Container = context.container;

  const render: preact.FunctionComponent<Props<ContainerProps, T>> = ({
    opened,
    dialogRoot,
    container,
    ...props
  }: Props<ContainerProps, T>) => {
    return createPortal(
      <Root aria-hidden={!opened} {...props}>
        <Overlay opened={opened} />
        {container && h(Container, { ...container, opened })}
      </Root>,
      dialogRoot
    );
  };
  render.displayName = "Dialog";

  return render;
};
