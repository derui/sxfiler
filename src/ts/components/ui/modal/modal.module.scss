@import "../commom-style/reset.module.css";

// base layout for root element. This layout should use without change.
.root-base {
  display: flex;
  position: absolute;
  width: 100%;
  height: 100%;
  justify-content: center;

  &[aria-hidden="true"] {
    display: none;
  }
}

// Base layout for overlay element.
.overlay-base {
  position: absolute;
  width: 100%;
  height: 100%;

  z-index: 1;
}

// Base layout for madal elemenrt.
.container-base {
  z-index: 2;
}
