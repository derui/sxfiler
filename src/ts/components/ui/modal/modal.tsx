import * as React from "react";
import * as ReactDOM from "react-dom";
import * as Element from "../element/element";
import { applyDisplayName } from "../util";

export interface ModalClassNames {
  root?: string;
  overlay?: string;
  container?: string;
}

export type Props<
  ContainerProps extends { className?: string } = Element.Props,
  OverlayProps extends { className?: string } = Element.Props,
  T extends Element.Props<H> = Element.Props,
  H extends HTMLElement = HTMLElement
> = T &
  React.HTMLAttributes<H> & {
    dialogRoot: HTMLElement;
    opened: boolean;
    container?: ContainerProps;
    overlay?: OverlayProps;
  };

export function createComponent<
  ContainerProps extends { className?: string } = Element.Props,
  OverlayProps extends { className?: string } = Element.Props,
  T extends Element.Props<H> = Element.Props,
  H extends HTMLElement = HTMLElement
>(
  context: {
    container?: React.ComponentType<ContainerProps & { opened: boolean; onClose: () => void; onOpen: () => void }>;
    overlay?: React.ComponentType<OverlayProps & { opened: boolean }>;
    classNames?: ModalClassNames;
  } = {}
): React.ComponentType<Props<ContainerProps, OverlayProps, T, H>> {
  const modalClassNames = context.classNames || {};
  const Container = context.container || Element.Component;
  const Overlay = context.overlay || Element.Component;

  return applyDisplayName("Dialog", ({ opened, dialogRoot, overlay, container, ...props }) => {
    const [closed, setClosed] = React.useState(true);

    return ReactDOM.createPortal(
      <Element.Component className={modalClassNames.root} aria-hidden={!opened && closed} {...props}>
        {overlay && <Overlay className={modalClassNames.overlay} opened={opened} {...overlay} />}
        {container && (
          <Container
            className={modalClassNames.container}
            onOpen={() => setClosed(false)}
            onClose={() => setClosed(true)}
            opened={opened}
            {...container}
          />
        )}
      </Element.Component>,
      dialogRoot
    );
  });
}

export const Component = createComponent();
