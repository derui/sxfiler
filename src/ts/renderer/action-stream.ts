import * as Rx from 'rx';
import {EventEmitter} from 'events';
import {RendererIPC} from './renderer-ipc';

type actionArgument = (ipc: RendererIPC) => Rx.Observable<any> | Rx.Observable<any>;

/**
 * An interface of ActionStream
 */
export interface ActionStream {
  /**
   * Publish action to action stream.
   *
   * @param {function(ipc:RendererIPC):Rx.Observable|Rx.Observable} act - an action to publish
   */
  doAction(act: actionArgument): void;

  /**
   * Subscribe events from this.
   *
   * @param {function(value: any)|Rx.Observer<any>} onNext - an onNext callback or observer
   * @param {function(exception: any)} onError - an onError callback
   * @param {function()} onCompleted - an onCompleted callback
   * @return {Rx.IDisposable} a disposable
   */
  subscribe(onNext?: ((value: any) => void|Rx.Observer<any>), onError?: (exception: any) => void, onCompleted?: () => void)
  : Rx.IDisposable;

  /**
   * Dispose published stream from this.
   */
  dispose(): void;
}

/**
 * ActionStream provide action emitter and method to subscribe it.
 */
export default class ActionStreamImpl implements ActionStream {
  private _e: NodeJS.EventEmitter;
  private _broadcastingObservable: Rx.Observable<any>;
  private _broadcastingDisposable: Rx.IDisposable;

  public constructor(ipc: RendererIPC) {
    this._e = new EventEmitter();

    // define observable from EventEmitter
    this._broadcastingObservable = Rx.Observable.fromEventPattern(
      (h: (data: any) => any) => this._e.addListener('action', h),
      (h: (data: any) => any) => this._e.removeListener('action', h),
      (act: any) => {
        if (typeof act === 'function') {
          return (<actionArgument>act)(ipc);
        } else {
          return act;
        }
      }
    ).flatMap((x: Rx.Observable<any>) => x).publish();

    this._broadcastingDisposable = this._broadcastingObservable.connect();
  }

  /**
   * Publish action to action stream.
   *
   * @param {function(ipc:RendererIPC):Rx.Observable|Rx.Observable} act - an action to publish
   */
  public doAction(act: actionArgument): void {
    this._e.emit('action', act);
  }

  /**
   * Subscribe events from this.
   *
   * @param {function(value: any)|Rx.Observer<any>} onNext - an onNext callback or observer
   * @param {function(exception: any)} onError - an onError callback
   * @param {function()} onCompleted - an onCompleted callback
   * @return {Rx.IDisposable} a disposable
   */
  public subscribe(onNext?: ((value: any) => void|Rx.Observer<any>), onError?: (exception: any) => void, onCompleted?: () => void)
  : Rx.IDisposable {
    return this._broadcastingObservable.subscribe(onNext, onError, onCompleted);
  }

  /**
   * Dispose published stream from this.
   */
  public dispose(): void {
    if (this._broadcastingDisposable) {
      this._broadcastingDisposable.dispose();
    }
    this._broadcastingDisposable = null;
  }
}
