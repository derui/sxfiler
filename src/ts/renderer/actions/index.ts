import * as directoryM from './directory';
import * as paneOperationM from './pane-operations';
import * as windowM from './window';

export const directory = directoryM;

export const paneOperation = paneOperationM;

export const window = windowM;
