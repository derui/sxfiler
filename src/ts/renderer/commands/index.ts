import * as Directory from './defaults/directory';
import * as Pane from './defaults/pane';
import * as Window from './defaults/window';

// export default commands.
import * as types from './types';

// exporting all types from current module.
export * from './types';

export const defaultCommands: types.Commands = {
  commands: [
    Directory.upDirectory,
    Directory.downDirectory,
    Pane.upCursor,
    Pane.downCursor,
    Pane.changePane,
    Window.quitApplication
  ]
};
