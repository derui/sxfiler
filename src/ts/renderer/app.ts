/// <reference path="../../../typings/tsd.d.ts" />
import * as Rx from 'rx';

import * as ReactDOM from 'react-dom';
import * as React from 'react';
import uuid from 'uuid';
import R from 'ramda';

import Main from './components/main';
import ActionStream from './action-stream';
import RendererIPC from './renderer-ipc';
import {PANE} from 'sxfiler/common/constants';
import Directory from './binders/directory';
import CurrentPane from './binders/current-pane';
import * as types from 'sxfiler/renderer/types';
import * as actions from './actions';
import KeyHandler from './key-handlers';
import {convertEventToDescriptor} from 'sxfiler/renderer/utils/key-binding-util';

let ipc = new RendererIPC(window.require('electron').ipcRenderer);
let actionStream = new ActionStream(ipc);

let currentPane = new CurrentPane(PANE.LEFT, actionStream);
let leftPane = new Directory(PANE.LEFT, actionStream);
let rightPane = new Directory(PANE.RIGHT, actionStream);
let paneInfo = Rx.Observable.combineLatest(
  leftPane.store.getObservable(),
  rightPane.store.getObservable(),
  (left: types.PaneInfo, right: types.PaneInfo): types.PaneInfos => {
    return {
      [PANE.LEFT]: left,
      [PANE.RIGHT]: right
    };
  });

let globalState = Rx.Observable.combineLatest(
  paneInfo,
  currentPane.store.getObservable(),
  (paneInfo: types.PaneInfos, currentPane: string): types.GlobalState => {
    return {
      paneInfo,
      currentPane
    };
  }).publish();

globalState.connect();

globalState.subscribe((state: types.GlobalState) => {
  ReactDOM.render(React.createElement(Main, {state, actionStream}), document.querySelector('.entry'));
});

// setup keyevent stream
interface StateWithEvent {
  event: KeyboardEvent;
  eventId: string;
  state: types.GlobalState;
};

let keyEvents = Rx.Observable.combineLatest(
  Rx.Observable.fromEvent(document.body, 'keydown')
    .select((e: KeyboardEvent) => {
      return {
        event: e,
        id: uuid.v4()
      };
    }),
  globalState,
  (ev: {event: KeyboardEvent, id: string}, state: types.GlobalState): StateWithEvent => {
    return {
      event: ev.event,
      eventId: ev.id,
      state
    };
  });

// this variable only store event id from observable sequence as `keyEvents'
let _lastEventId = '';
let keyHandler = new KeyHandler();
keyEvents.subscribe((state: StateWithEvent) => {
  if (state.eventId === _lastEventId) {
    return;
  }

  _lastEventId = state.eventId;

  let _e = R.clone(state.event);
  let desc = convertEventToDescriptor(_e);
  actionStream.doAction(
    keyHandler.invokeCommand(desc, state.state)
  );
});

actionStream.doAction(actions.directory.getDirectory('.', PANE.LEFT));
actionStream.doAction(actions.directory.getDirectory('.', PANE.RIGHT));
