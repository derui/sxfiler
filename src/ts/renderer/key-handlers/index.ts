// provide key handler with key bindings
import * as Rx from 'rx';
import R from 'ramda';
import * as types from 'sxfiler/common/types';
import {GlobalState} from 'sxfiler/renderer/types';
import * as commands from 'sxfiler/renderer/commands';
import {getKeyIdentifier, bindMappingWithCommands} from 'sxfiler/renderer/utils/key-binding-util';
import defaultMapping from './defaults';
import * as actions from 'sxfiler/renderer/actions';
import * as Maybe from 'sxfiler/common/maybe';

let defaultCommands = commands.defaultCommands;

/**
 * Merge two keyDescriptors to one keyDescriptors.
 *
 * @param {array[KeyDescriptor]} src1 key bindings to merge with src2
 * @param {array[KeyDescriptor]} src2 key bindings to merge with src1
 * @return {array[KeyDescriptor]} merged key descriptors
 */
function mergeDescriptors(src1: types.KeyDescriptor[], src2: types.KeyDescriptor[]): types.KeyDescriptor[] {
  'use strict';

  let toObj: any = R.reduce((memo: any, v: types.KeyDescriptor) => R.merge(memo, {[getKeyIdentifier(v)]: v}), {});
  let src1Descs = toObj(src1), src2Descs = toObj(src2);

  return R.values(R.merge(src1Descs, src2Descs));
}

export interface KeyHandler {
  /**
   * Invoke a command binded a key-combination which calculated from KeyboardEvent.
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @param {IGlobalState} globalState - the global state to pass command
   * @return {Rx.Observable} result of command
   */
  invokeCommand(desc: types.KeyDescriptor, globalState: GlobalState): Rx.Observable<any>;
}

/**
 * KeyHandler provide match keyboard input with keybindings that are defined user or default,
 *  and run command in it.
 */
export default class KeyHandlerImpl implements KeyHandler {
  private keyBindings: types.KeyBindings<commands.Command>;

  /**
   * Construct KeyHandler.
   *
   * @param {IKeyBindings} [keyBindings] key bindings to append to default binding
   */
  constructor(keyBindings?: types.KeyBindings<commands.Command>) {
    this.keyBindings = bindMappingWithCommands(defaultMapping, defaultCommands);

    // if keybindings given, merge with default bindings.
    if (keyBindings) {
      this.keyBindings.bindings = mergeDescriptors(this.keyBindings.bindings, keyBindings.bindings);
      this.keyBindings.mapping = R.merge(this.keyBindings.bindings, keyBindings.mapping);
    }
  }

  /**
   * Invoke a command binded a key-combination which calculated from KeyboardEvent.
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @param {IGlobalState} globalState - the global state to pass command
   * @return {Rx.Observable} result of command
   */
  public invokeCommand(desc: types.KeyDescriptor, globalState: GlobalState): Rx.Observable<any> {

    let command = this.matchKey(desc);
    if (Maybe.isNone(command)) {
      return Rx.Observable.empty();
    }

    return command.val.command(globalState, actions);
  }

  /**
   * Match given KeyboardEvent with keybindings in this class
   *
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @return {ICommand} a command if match keyboard event with keybindings. return null if not.
   */
  private matchKey(desc: types.KeyDescriptor): Maybe.Data<commands.Command> {
    if (!desc) {
      return Maybe.NONE;
    }

    let id = getKeyIdentifier(desc);

    let binding = this.keyBindings.mapping[id];
    if (!binding) {
      return Maybe.NONE;
    }
    return Maybe.just(binding);
  }
}
