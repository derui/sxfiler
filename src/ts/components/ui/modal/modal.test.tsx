import * as React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import * as Modal from "./modal";

describe("UI kit", () => {
  describe("Modal", () => {
    it("render with empty nodes", () => {
      const C = Modal.createComponent();
      const root = document.createElement("div");
      const tree = renderer
        .create(
          <C dialogRoot={root} opened={false} container={{ role: "dialog" }} overlay={{ "aria-label": "overlay" }} />
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("render with component parameters ", () => {
      const C = Modal.Component;
      const root = mount(<div />);
      const tree = mount(
        <C
          dialogRoot={root.getDOMNode() as HTMLElement}
          opened={true}
          container={{ role: "dialog" }}
          overlay={{ "aria-label": "overlay" }}
        />
      );

      expect(tree.exists("[aria-label='overlay']")).toBeTruthy;
    });

    it("render a dialog opened on a dialog root", () => {
      const C = Modal.createComponent({
        classNames: { root: "foo" },
      });
      const root = mount(<div />);
      mount(<C dialogRoot={root.getDOMNode() as HTMLElement} opened={true} />);

      expect(root.exists(".foo")).toBeTruthy;
    });
  });
});
