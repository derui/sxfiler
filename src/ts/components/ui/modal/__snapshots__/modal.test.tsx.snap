// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI kit Modal render with empty nodes 1`] = `null`;
